Add how to register a sidechain page in lisk docs (#1653)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/register-sidechain.adoc

Each item contains the following properties:

. `blsKey`: The {url_bls_key}[public BLS key] of the validator.
. `bftWeight`: The {url_lip56}[BFT weight^] is the weight attributed to the pre-votes and pre-commits cast by a validator, and therefore determines to what extent the validator contributes to finalizing blocks.

To retrieve the BFT parameters of all sidechain validators, invoke the `consensus_getBFTParameters` endpoint on the *sidechain node*.

Set the `height` to the *latest block height on the sidechain* in the request parameters.

TIP: The xref:{url_api_rpc_getnodeinfo}[system_getNodeInfo] node endpoint can be requested, to retrieve the latest block height.

[tabs]
=====
Sidechain node CLI::
+
--
[source,bash]
----
./bin/run  endpoint:invoke consensus_getBFTParameters '{"height": 200}'
----
--
cURL::
+
--
[source,bash]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "consensus_getBFTParameters",
    "params": {
         "height": 200
    }
}'
----
--
=====

> This will return a complete list of all active sidechain validators at the specified block height.

Reviewer: good catch, would have been too easy the other way :D I have added a link to the script and adjusted the content

* If the mainchain and/or sidechain registration command was submitted with the wrong data, then the whole sidechain needs to be restarted completely with the correct values, or it needs to hardfork in order to connect to the mainchain.
====

To create a *Register Mainchain* transaction, the following information is required:

. <<preparing-the-register-mainchain-parameters,Mainchain registration parameters>>
.. <<ownchainid>>
.. <<ownname>>
.. <<mainchainvalidators>>
.. <<mainchaincertificatethreshold>>
. <<preparing-the-aggregated-signature,The aggregated signature of the sidechain validators>>

=== Preparing the Register Mainchain parameters

The Register Mainchain parameters are created in an analog way as the Register Sidechain parameters have been created in the previous step.

To conveniently prepare the required parameters, check out the following script in the lisk-sdk-examples repository:
{url_exmaples_guides_createparams}[paramsCreation.ts^]

The script is executed like this:

[source,bash]
----

> node paramsCreation.ts

Reviewer: what exactly are the changes and why did you do them?

> node paramsCreation.ts

Reviewer 2: Defined constants in the same file, added lines for both IPC and WS client etc.

> node paramsCreation.ts

Reviewer: Ok, I updated the script accordingly 👍

By signing the mainchain registration message, they verify the correctness of the following values: `ownChainID`, `ownChainName`, `mainchainCertificateThreshold` and `mainchainValidators`.
All individual signatures are then aggregated into one `signature` and the corresponding `aggregationBits` which are appended to the Register Mainchain parameters.

The individual signing of the mainchain registration by the sidechain validators is executed *off-chain*, similar to the process of creating multi-signature transactions.

==== Collecting the signatures of the sidechain validators

The process of signing the registration parameters needs to be coordinated off-chain by the sidechain validators.

To simplify the process, you may use the following script, which adds a validator signature to an existing list of signatures: {url_exmaples_guides_validatorsignatures}[validatorSignatures.ts]

The script expects the following files to be present:

. `params.json`, which was created in step <<preparing-the-register-mainchain-parameters>>.
. `keys.json`: a file with the bls keys of the validator intending to sign.

> This file can be generated for a specific validator with the `keys:create` CLI command, by providing the passphrase.

Reviewer: Mentioning the count flag is not really relevant at this point, because the default and recommended scenario is that a user manages one validator account, not many

> This file can be generated for a specific validator with the `keys:create` CLI command, by providing the passphrase.

Reviewer 2: Alright, I think we should at least mention the command about how they can create the `keys.json` to aid the reader. Alternatively, we can refer to the CLI command reference page as well:

https://liskhq.github.io/lisk-docs/lisk-sdk/v6/client-cli.html#keyscreate

> This file can be generated for a specific validator with the `keys:create` CLI command, by providing the passphrase.

Reviewer: I have added the link 👍

. `sidechainValidatorSignatures.json`: List of already existing validator signatures.
If no such file is found, it will be created newly.

After installing the required dependencies, the script can be executed like so:

[source,bash]
----

> node validatorSignatures.ts

Reviewer: This is intentional, because it is not a realistic scenario that one user creates all the signatures of the sidechain validators.

> node validatorSignatures.ts

Reviewer 2: IMHO we should either go with one validator on both the sidechain and mainchain registration process or use the list of validators we create on lines 166-176. Having a single validator here will lead to an error while trying out the guide. I think even with one validator, we can send a Register Sidechain and later a Register Mainchain transaction so, it should be in sync.

> node validatorSignatures.ts

Reviewer: As this process happens in a decentralized environment, it is in fact not possible that a single person tried out this whole guide successfully, except of a development network, which is centralized. And there is no point to explain the artificial scenario of how to register a sidechain in a centralized network.

The normal scenario is that it is a decentralized network, and the process of mainchain registration is therefore decentralized as well, this is what the guide should explain and emphasize. If, after reading the guide, this particular part is still unclear to you, maybe we should improve the section explaining this particular topic.

> node validatorSignatures.ts

Reviewer: From perspective of a normal mainchain node operator, there is nothing special to do. Also, not every sidechain node operator needs to do anything, only sidechain validators need to take action here. So not really sure what to improve, but if you agree we can improve it later in another issue :)

> node validatorSignatures.ts

Reviewer 2: Alright, thanks :)

const signature = bls.signData(
      MESSAGE_TAG_CHAIN_REG,
      params.ownChainID,
      message,
      Buffer.from(blsKeys[0].plain.blsPrivateKey,"hex"),
    );
----

The validator signatures need to be collected by the person intending to send the Register Mainchain transaction.

Once the aggregated BFT weight of the validators who signed is equal or above the <<sidechaincertificatethreshold>>, enough validators have signed the registration parameters.

==== Aggregating the validator signatures

Once enough validators added their signatures, the list of signatures is aggregated into one single signature, which is then appended to the registration params that we created in step <<preparing-the-register-mainchain-parameters>>.

The following script is performing the required steps: {url_exmaples_guides_mcregistration}[mainchainRegistration.ts]

The script expects the following files to be present:

. `params.json`, which was created in step <<preparing-the-register-mainchain-parameters>>.
. `sidechainValidatorSignatures.json`: List of validator signatures.

After installing the required dependencies, the script can be executed like so:

[source,bash]
----

> node mainchainRegistration.ts

Reviewer: I have removed the broken verification for now, maybe we can revisit this later

> node mainchainRegistration.ts

Reviewer 2: Thanks, to aid the reader further, I guess we can leave a TIP somewhere with the aforementioned link suggesting that if they want to create the mainchain params using a single script, they can use this [file](https://lightcurve.atlassian.net/wiki/pages/viewpage.action?pageId=6127961&pageVersion=37&preview=/6127961/8094237/prepare_mainchain_reg_params.ts):

> node mainchainRegistration.ts

Reviewer: I don't think this file is public. But I created a new issue to add a signature verification at a later point here: https://github.com/LiskHQ/lisk-sdk-examples/issues/308

Once the sidechain registration command is processed, the sidechain account status is set to registered.
In this state, the cross-chain channel is still not active, so the users on the mainchain or other chains cannot send xref:{url_ccm}[cross-chain messages], (CCMs) to this sidechain yet.
Moreover, the liveness requirement to maintain the channel is not enforced, this means that there is no specific time requirement for a sidechain to be activated on the mainchain, it can stay in the registered status for any period of time.
When a first valid cross-chain update command from this sidechain is processed, the sidechain status is changed to active, making it active in the ecosystem.
Now it is possible to send CCMs to the sidechain and the liveness condition is enforced.

=== Connecting a Sidechain

The sidechain registration command is used to register a sidechain on the Lisk mainchain.
Once this command is processed, a new account for the sidechain is created in the mainchain state under the interoperability store.
The account is initialized with an empty inbox and outbox, while the sidechain name and the initial validators set are given in the command parameters.

> The network ID is calculated from the address of the command sender and the genesis block ID, also given in the command parameters.

Reviewer 3: This is slightly outdated, now the sidechain has the liberty of choosing the chain ID itself. Also the term "network ID" is not used anymore, it is replaced by "chain ID".

Author: OK thanks, I have reworded this 👍

> The following parameters below are included in the registration command:

Reviewer 3: I think you refer to the outdated version of the command. The current arguments could be found here: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0043.md#sidechain-registration-command.
The differences are: command sender explicitly passes `chainID`, `certificateThreshold` is a separate parameter.

Author: Hi Sergey, As discussed, I have restructure this. 👍

* `name` -  The name property sets the name of the sidechain as a string of characters, and has to be unique within the Lisk ecosystem.

// * `genesisBlockid` - The ID of the genesis block ID  (as is defined in xref:{url_lip34}[LIP34]), is computed from the xref:{url_sha_256}[SHA-256] digest of the serialized bytes of the sidechain genesis block.
// It can also help future sidechain node operators to identify the sidechain genesis block with respect to its value.

* `initValidators` - This property defines the set of eligible keys with their respective weights and the certificate threshold required to sign the first certificate from the sidechain.
It is an object containing the following properties:

.. `keys`: An array of xref:{url_bls_key}[BLS public keys].
The set of public keys that are eligible to sign the next certificate.

.. xref:{url_lip56}[weights]: An array of integers of the same size as the keys property, where each element is the weight of the corresponding key.

> For PoS chains, the value of the elements of this array is usually 1 as every active validator has the same finality weight to sign the next certificate.

Reviewer 3: Now this parameter is called `bftWeight`. For PoS chains it should correspond to the stake of the validator, not be constant 1, this is also outdated with our latest changes in DPoS (PoS) module.

Author: Thanks, I have updated this paragraph 👍

.. `certificateThreshold`: An integer setting the minimum signatures weight required for the first sidechain certificate to be valid.

=== Chain Identifiers

The chain identifiers for both transaction and block signatures were previously known as network identifiers.
The chain ID can be considered as a development of the network ID concept.
It is used to prevent transaction replay attacks in different networks, and now also helps to identify the chain in the ecosystem.

Chain identifiers are 4-byte values that follow a specific format: the first byte is used to identify the network in which the chain is running (either the mainnet, official testnet, or any other test network); the other 3 bytes identify the chain within the network.
The network-specific prefix is included explicitly to ensure that a chain does not use the same chain identifier in the test network as in the mainnet.

> By using 4 bytes instead of 32 bytes this provides the distinct advantage whereby users are able to easily verify that they are signing a transaction for the correct blockchain.

Reviewer 3: I don't quite understand this or the following sentence. I think, we never thought about 32 byte values for chain ID (maybe you confused with 32 bits = 4 bytes?).
I would just mention that by checking chain ID users are able to easily verify that they are signing a transaction for the correct blockchain.

In addition, the chain identifier can be directly set by the blockchain creator, which is far more convenient than generating a random 32-byte value.

The first byte is set to `CHAIN_ID_PREFIX_MAINNET` for chains running in the mainnet network and to `CHAIN_ID_PREFIX_TESTNET` for chains running in the testnet network.
The other 3 bytes must be chosen uniquely for the respective blockchain, i.e., no other blockchain created with the Lisk SDK should use the same 3 bytes.

The chain-identifiers prefixes currently specified can be found here in the xref:{url_reg_recovery}[Sidechain registration and recovery page].

> === Chain ID

Reviewer 3: As I mentioned before, the chain ID is now not deterministically generated but rather chosen by the sidechain itself when submitting a sidechain reg command. See here: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0043.md#chainid.

    "required": [
        "name",
        "chainID",
        "initValidators",
        "certificateThreshold"
    ],
    "properties": {
        "name": {
            "dataType": "string",
            "minLength": MIN_CHAIN_NAME_LENGTH,
            "maxLength": MAX_CHAIN_NAME_LENGTH,
            "fieldNumber": 1
        },
        "chainID": {
            "dataType": "bytes",
            "length": CHAIN_ID_LENGTH,
            "fieldNumber": 2
        },
        "initValidators": {
            "type": "array",
            "fieldNumber": 3,
            "items": {
                "type": "object",
                "required": ["blsKey", "bftWeight"],
                "properties": {
                    "blsKey": {
                        "dataType": "bytes",
                        "length": BLS_PUBLIC_KEY_LENGTH,
                        "fieldNumber": 1
                    },
                    "bftWeight": {
                        "dataType": "uint64",
                        "fieldNumber": 2
                    }
                }
            }
        },
        "certificateThreshold": {
            "dataType": "uint64",
            "fieldNumber": 4
        }
    }
}
----

> All the sidechain registration, verification, and execution parameters can be found  xref:{url_sidechain_reg_commands}[here].

Reviewer 3: The link doesn't work.

Author: It should work now.

== Mainchain registration

A similar registration process has to be initiated within the sidechain to enable an interoperable channel to fuction.
This is achieved by performing a transaction with the xref:{url_mainchain_reg}[Mainchain Registration Command] within the respective sidechain.

> One of the key differences here between the sidechain registration command, is to activate the mainchain registration command, this requires that a substantial majority of the current active sidechain validators approve and sign this transaction.

Reviewer 3: This sentence reads weird, I would try to reformulate (the content is correct).

Author: Hmmm..OK, I have tried to reformulate it and make it somewhat more direct, please let me know your thoughts 👍

            "dataType": "bytes",
            "length": CHAIN_ID_LENGTH,
            "fieldNumber": 1
        },
        "ownName": {
            "dataType": "string",
            "minLength": MIN_CHAIN_NAME_LENGTH,
            "maxLength": MAX_CHAIN_NAME_LENGTH,
            "fieldNumber": 2
        },
        "mainchainValidators": {
            "type": "array",
            "fieldNumber": 3,
            "items": {
                "type": "object",
                "required": ["blsKey", "bftWeight"],
                "properties": {
                    "blsKey": {
                        "dataType": "bytes",
                        "length": BLS_PUBLIC_KEY_LENGTH,
                        "fieldNumber": 1
                    },
                    "bftWeight": {
                        "dataType": "uint64",
                        "fieldNumber": 2
                    }
                }
            }
        },
        "signature": {
            "dataType": "bytes",
            "length": BLS_SIGNATURE_LENGTH,
            "fieldNumber": 4
        },
        "aggregationBits": {
            "dataType": "bytes",
            "fieldNumber": 5
        }
    }
}
----

> All the mainchain registration, verification, and execution parameters can be found  xref:{url_mainchain_reg_commands}[here].

Reviewer 3: The link doesn't work.

Reviewer 3: Hm, previous link was fixed, but this one still doesn't work.

Author: Try now, I had a lot of issues with my new version of Webstorm 👍

These correspond to the three stages of the sidechain lifecycle.
The sidechain account is automatically set to “registered” when it is created.
As previously covered in the xref:{url_sidechain_reg_recovery}[Sidechain registration and recovery page], a sidechain can be defined as a separate blockchain next to the mainchain, and the inclusion of a transaction in one of the blockchains can have an effect on the other blockchain.
In this section, the registration process is explained.

== Registration overview

The interoperability module is responsible for storing and validating the properties necessary to process the registration commands.
In this section, we will focus on both the sidechain and mainchain registration requirements necessary when registering a sidechain.
Each interoperable sidechain maintains a chain account, channel, and the chain validators for the mainchain.
In addition, the mainchain maintains these aforementioned data structures for each registered sidechain.

== Sidechain registration procedure

A sidechain registration command can be sent by any user account in the Lisk Mainchain that has adequate funds to pay the required fee.

> The processing of this command is designed to create a sidechain account in the mainchain state associated with a unique network identifier and a name.

Reviewer 3: Here also the term "network identifier" should be changed to "chain identifier".

Author: Ah yes...Well spotted Sergey! Many thanks, it is now corrected 👍

:url_mainchain_reg: understand-blockchain/interoperability/sidechain-registration-and-recovery.adoc#mainchain-registration-command
:url_mainchain_reg_commands: understand-blockchain/interoperability/sidechain-registration-and-recovery.adoc#mainchain-registration-on-a-sidechain
:url_chain_id: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0043.md
// Explain how to register a sidechain to a mainchain in build blockchain section.
// Explain command to register a sidechain
// Explain how to define chainID with concept of networkID
// Explain how to register mainchain to a sidechain

== Lisk Sidechains

Within the Lisk ecosystem a sidechain can be in one of three status modes: “registered”, “active”, and “terminated”.
These correspond to the three stages of the sidechain lifecycle.
The sidechain account is automatically set to “registered” when it is created.
As previously covered in the xref:{url_sidechain_reg_recovery}[Sidechain registration and recovery page], a sidechain can be defined as a separate blockchain next to the mainchain, and the inclusion of a transaction in one of the blockchains can have an effect on the other blockchain.

> In this section, the registration process is explained.

Reviewer: We already explain what is a sidechain in the understand blockchain category, so we can asume here the reader knows what a sidechain is.

Please improve the introduction section and explain what the reader will learn in this guide.

Author: Yes, good point Mona, I have restructured the introduction section 👍

== Registration overview

The interoperability module is responsible for storing and validating the properties necessary to process the registration commands.
In this section, we will focus on both the sidechain and mainchain registration requirements necessary when registering a sidechain.
Each interoperable sidechain maintains a chain account, channel, and the chain validators for the mainchain.

> In addition, the mainchain maintains these aforementioned data structures for each registered sidechain.

Reviewer: Please add an overview of all required steps here, e.g. as a numbered list.

Author: OK, I have restructured the first 2 sections, which does now give a numbered overview list.

Set the `height` to the *latest block height on the sidechain* in the request parameters.

TIP: The `system_getNodeInfo` node endpoint can be requested, to retrieve the latest block height.

[tabs]
=====
Sidechain node CLI::
+
--
[source,bash]
----
./bin/run  endpoint:invoke consensus_getBFTParameters '{"height": 200}'
----
--
cURL::
+
--
[source,bash]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "consensus_getBFTParameters",
    "params": {
         "height": 200
    }
}'
----
--
=====

This will return a complete list of all active sidechain validators at the specified block height.

Now, paste the *complete list* that was returned by the `consensus_getBFTParameters` endpoint under the `"sidechainValidators"` key:

.registerSidechain.json
[source,json]
----
{
  "chainID": "00000001",
  "name": "sidechain1",
  "sidechainValidators": [
	{
	  "blsKey": "92f020ce5e37befb86493a82686b0eedddb264350b0873cf1eeaa1fefe39d938f05f272452c1ef5e6ceb4d9b23687e31",
	  "bftWeight": "1"
	},
   // ...

>   ]

Reviewer: I think `certificateThreshold` is just a valid default value for the threshold. About the other properties, I am not sure if they are relevant in this context, maybe @ishantiw can clarify :)

Minimum and maximum values for the certificate threshold are calculated with the <<aggregated-bft-weight>>.

.Calculating the aggregated BFT weight for sidechain validators
[source,js]
----
const { apiClient } = require('@liskhq/lisk-client');

(async () => {
	// Update the path to point to your sidechain app data folder
	const sidechainClient = await apiClient.createIPCClient('~/.lisk/hello_client');
	const sidechainNodeInfo = await sidechainClient.invoke('system_getNodeInfo');
	// Get active validators from sidechain
	const bftParams = await sidechainClient.invoke('consensus_getBFTParameters', { height: sidechainNodeInfo.height });

	// Calculate the aggregated BFT weight
	let aggregateBFTWeight = BigInt(0);
	for (const validator of bftParams.validators) {
		aggregateBFTWeight += BigInt(validator.bftWeight);
	}

	console.log("certificateThreshold:");
	console.log("min:");
	console.log(aggregateBFTWeight/BigInt(3) + BigInt(1));
	console.log("max:");
	console.log(aggregateBFTWeight);
	process.exit(0);
})();
----

For this example, the minimum value is 34, and the maximum 101.

> Taking this into account, we decide for a `sidechainCertificateThreshold` of 65:

Reviewer: I think the `certificateThreshold` is a valid default value, in the middle between the minimum and maximum possible values, But maybe @ishantiw can clarify.

65 is just chosen here in the example, it could be any value between 34 and 101, which is mentioned in the previous sentence.

}
----

Copy the transaction and send it to the node, for example by using the node CLI:

[source,bash]
----

> ./bin/run transaction:send 0a10696e7465726f7065726162696c6974791211726567697374657253696465636861696e18002080a8d6b9072a20a3f96c50d0446220ef2f98240898515cbba8155730679ca35326d98dcfb680f0324a0a0404000001120a73696465636861696e311a340a3092f020ce5e37befb86493a82686b0eedddb264350b0873cf1eeaa1fefe39d938f05f272452c1ef5e6ceb4d9b23687e31100220023a408261e374405af4ec1143dfc0ae82a38e385d0edce870f698385749112064b374ac0de67354210aa27280db82121ec0bce195e5630c56a568a8b99dbbcb3a3d0b

Reviewer: yes :)

. `keys.json`: a file with the bls keys of the validator intending to sign.
This file can be generated for a specific validator with the `keys:create` CLI command, by providing the passphrase.
. `sidechainValidatorSignatures.json`: List of already existing validator signatures.
If no such file is found, it will be created newly.

After installing the required dependencies, the script can be executed like so:

[source,bash]
----
node validatorSignatures.ts
----

As a result, it will append a new signature to the signatures listed in the file `sidechainValidatorSignatures.json`.

The most important parts of the script are the following:

[source,js]
----
// 1. Preparing params for the registration message
params = {
    ownChainID: Buffer.from(paramsJSON.ownChainID, 'hex'),
    ownName: paramsJSON.ownName,
    mainchainValidators: paramsJSON.mainchainValidators.map(v => ({
        blsKey: Buffer.from(v.blsKey, 'hex'),
        bftWeight: BigInt(v.bftWeight),
    })),
    mainchainCertificateThreshold: paramsJSON.mainchainCertificateThreshold.toString(),
};

// 2. Create message by encoding params

> const message = codec.codec.encode(registrationSignatureMessageSchema, params);

Reviewer 2: ```suggestion
const message = codec.encode(registrationSignatureMessageSchema, params);
```

Reviewer: i think this snippet is actually correct, the way codec is imported

----

If the node accepted the transaction, it will respond with the transaction ID.

----
Transaction with id: '1d944f3fb46714978ad7bedd1b788919c3b37e92d893088fd056f8217f20ed8a' received by node.
----

== How to verify the mainchain registration
=== Check the mainchain account
Once the "Register Mainchain" command is processed, the mainchain account is initialized and its `status` is set to `registered`.

To verify that the account was created successfully, request the `interoperability_getChainData` endpoint from a sidechain node.

Parameters:

* `chainID`: The chain ID of the registered mainchain.

[tabs]
=====
Node CLI::
+
--
If you maintain an own instance of a sidechain node, it is possible to invoke endpoints directly via the node CLI:

[source,bash]
----

> ./bin/run endpoint:invoke interoperability_getChainData '{"chainID": "00000000"}'

Reviewer 2: Endpoint's name might need an update!

Reviewer: I think this is correct considering the latest updates, maybe @ishantiw can clarify

Reviewer: Yeah but @ishantiw pointed outthat this endpoint will change to `interoperability_getChainData`

            "validatorsHash": "58fa1be3fca7aef9952a7640397124229837079b14a144907a7e3373685daceb"
        },
        "name": "sidechain_8",
        "status": 0
    }
}
----

=== Check the sidechain channel

To verify that the channel to the sidechain was created successfully, request the `interoperability_getChainAccount` endpoint from a mainchain node.

Parameters:

* `chainID`: The chain ID of the registered sidechain.

[tabs]
=====
Mainchain node CLI::
+
--
[source,bash]
----
lisk-core endpoint:invoke interoperability_getChannel '{"chainID": "00000001"}'
----
--
cURL::
+
--
[source,bash]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "interoperability_getChannel",
    "params": {
         "chainID": "00000001"
    }
}'

----
--
=====

This will return the respective sidechain account stored in the xref:{url_sidechain_channel_store}[Channel substore] of the mainchain.

.A newly created sidechain channel on the mainchain
[source,json]
----
{
    "id": "1",
    "jsonrpc": "2.0",
    "result": {

>         "messageFeeTokenID": "0300000000000000",

Reviewer 3: I am slightly confused about this value. This corresponds to LSK token ID in network with id `03`, but you have registered a sidechain with network ID `00`. I think this should not be possible. Could you please double check this example with SDK?

Reviewer: Good catch, I tested the process on the alphanet, but wanted the guide examples to refer to mainnet and forgot to update this value. Should be fixed now 👍

.A newly created chain validators list on the mainchain
[source,json]
----
{
    "id": "1",
    "jsonrpc": "2.0",
    "result": {
        "activeValidators": [
            {
                "blsKey": "8012798d2ac6b93df3bfa931192d9b2d496c3c947958a7408232e08872895557c06c1b94f5cc2555e28addbaadf3e0bd",
                "bftWeight": "1"
            },
            // ...
        ],
        "certificateThreshold": "66"
    }
}
----

== How to register the mainchain on the sidechain

The mainchain is registered on a sidechain by posting a xref:{url_mainchain_reg}[Register Mainchain transaction].
A "Register Mainchain" transaction can be sent by any user account in the sidechain that has adequate funds to pay the required fee.

[IMPORTANT]
====
* The mainchain registration process always has to occur *after* the sidechain registration on the mainchain, since the sidechain has no prior knowledge of its name and must be certain that the correct chain ID has been registered.
* Connections to both mainchain and sidechain node APIs are required to create a Register Mainchain transaction.
====

[CAUTION]
====

> * It is of key importance that the sidechain validators ensure that they are signing the registration command with the correct information, otherwise the sidechain interoperable functionality may be unusable.

Reviewer 3: Maybe you could add that if the mainchain registration command was submitted with the wrong data, then the whole sidechain needs to be restarted. I am pretty sure that someone at some point will mess up and will look for solutions how to fix it :)

Reviewer: Just to clarify, what exactly do you mean with "it needs to be restarted"?

====

// call RPC to check if chain id available for the registered and it should return false
// `interoperability_isChainIDAvailable` with params "params": { "chainID": "04000001" } return false

.sidechain_reg_params.json
[source,json]
----
{
  "chainID": "00000001"
}
----

==== Name
The name of the sidechain as a string.
Has to be *unique* within the Lisk ecosystem  and contain only characters from the set `[a-z0-9!@$&_.]`.

.sidechain_reg_params.json
[source,json]
----
{
  "chainID": "00000001",
  "name": "sidechain1"
}
----

==== sidechainValidators
Defines the set of sidechain validators expected to sign the first certificate from the sidechain.
Each item contains the following properties:

. `blsKey`: The {url_bls_key}[public BLS key] of the validator.
. `bftWeight`: The {url_lip56}[BFT weight^] is the weight attributed to the pre-votes and pre-commits cast by a validator, and therefore determines to what extent the validator contributes to finalizing blocks.

To retrieve the BFT parameters of all sidechain validators, invoke the `consensus_getBFTParameters` endpoint on the *sidechain node*.

> Set the `height` to the *latest block height on the sidechain* in the request parameters:

Reviewer 4: here you can also mention to retrieve the latest height from `system_getNodeInfo` RPC as some users might want to automate this

}
----

> Copy the transaction and send it ot the node, for example by using the node CLI like so:

Reviewer 4: ```suggestion
Copy the transaction and send it to the node, for example by using the node CLI:
```

----

If the node accepted the transaction, it will respond with the transaction ID.

----
Transaction with id: '1d944f3fb46714978ad7bedd1b788919c3b37e92d893088fd056f8217f20ed8a' received by node.
----

== How to verify the sidechain registration

=== Check the sidechain account
Once the Register Sidechain command is processed, the sidechain account `status` is set to `registered`.

To verify that the account was created successfully, request the `interoperability_getChainAccount` endpoint from a mainchain node.

Parameters:

* `chainID`: The chain ID of the registered sidechain.

[tabs]
=====
Mainchain node CLI::
+
--
[source,bash]
----
lisk-core endpoint:invoke interoperability_getChainAccount '{"chainID": "00000001"}'
----
--
cURL::
+
--
[source,bash]
----
curl --location --request POST 'http://localhost:7887/rpc' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": "1",

>     "method": "interoperability_getChainAccount",

Reviewer 4: FYI: we are going to change this RPC to `interoperability_getChainData` to align with the LIP

////
In this example, we decide for a `mainchainCertificateThreshold` of 55:

.registerMainchain.json
[source,json]
----
{
  "ownChainID": "00000001",
  "ownName": "sidechain1",
  "mainchainValidators": [
	{
	  "blsKey": "92f020ce5e37befb86493a82686b0eedddb264350b0873cf1eeaa1fefe39d938f05f272452c1ef5e6ceb4d9b23687e31",
	  "bftWeight": "1"
	},
    // ...
  ],
  "mainchainCertificateThreshold": "55"
}
----
////

=== Preparing the aggregated signature
To create a valid signature, an amount of sidechain validators equal or greater to the <<sidechaincertificatethreshold>> need to individually sign the Mainchain registration with their account.
All individual signatures are then aggregated into one signature which is appended to the Register Mainchain parameters.

The individual signing of the mainchain registration by the sidechain validators is executed *off-chain*, similar to the process of creating multi-signature transactions.

==== Collecting the signatures of the sidechain validators

The process of signing the registration parameters needs to be coordinated off-chain by the sidechain validators.

> To simplify the process, you may use the following script, which adds a validator signature to an existing list of signatures: {url_exmaples_guides_validatorsignatures}[validatorSignatures.ts]

Reviewer: Why for testing only? How should it be done in a real scenario?

> To simplify the process, you may use the following script, which adds a validator signature to an existing list of signatures: {url_exmaples_guides_validatorsignatures}[validatorSignatures.ts]

Reviewer 4: I checked the `validatorSignatures.ts` and saw you commented the loop part for devnet validators. So i think this is correct. You can maybe simplify it by saying that individual validators can sign as below,

```ts
params = {
        ownChainID: Buffer.from(paramsJSON.ownChainID, 'hex'),
        ownName: paramsJSON.ownName,
        mainchainValidators: paramsJSON.mainchainValidators.map(v => ({
            blsKey: Buffer.from(v.blsKey, 'hex'),
            bftWeight: BigInt(v.bftWeight),
        })),
        mainchainCertificateThreshold: paramsJSON.mainchainCertificateThreshold.toString(),
    };

    // Create message by encoding params
    const message = codec.codec.encode(registrationSignatureMessageSchema, params);
    
    const signature = bls.signData(
          MESSAGE_TAG_CHAIN_REG,
          params.ownChainID,
          message,
          Buffer.from(blsKeys[0].plain.blsPrivateKey,"hex"),
        );
```

Maybe refine it a bit but they can simply share the signature to the person who is sending the registration transaction
